DEB-8 Add patient registration

MattyJ007 opened the pull request.

MattyJ007 committed:
Add patient registration OpenHIM Channel

Needed to route patient resources via the OpenHIM to the mapping
mediator

DEB-25

MattyJ007 committed:
Temp

bradsawadye commented on docker/importer/volume/openhim-import.json:
```diff
@@ -1,52 +1,61 @@
+      "requestBody": true,
+      "responseBody": true,
+      "rewriteUrlsConfig": [],
+      "name": "Patient Registration for Coivd19 Immunization",
```
```suggestion
      "name": "Patient Registration for Covid19 Immunization",
```

bradsawadye committed:
fix typo

MattyJ007 committed:
Add example Patient Resource message to readme

Quick start for users to test this package.

DEB-20

MattyJ007 committed:
Merge branch 'DEB-8-patient-registration' of github.com:jembi/covid19-immunization-tracking-package into DEB-8-patient-registration

MattyJ007 committed:
Add client to this packages dependencies

The immunization flow uses OpenCR to resolve the patient. Therefore it
needs the client package started up

DEb-20

MattyJ007 committed:
Fix a few package names (container and channel)

The importer container name was a duplicate of another package

The OpenHIM config referenced localhost as the host which does not work
with Docker
Making the OpenHIM Channel path name and endpoint name the same makes
data trafficking easier

DEB-20

MattyJ007 committed:
Update the OpenHIM mapping mediator version

The latest minor release contains needed features for this flow
Object pass through

DEB-20

MattyJ007 committed:
Add temporary working patient registration flow

This flow can now create a patient in HAPI-FHIR.
This flow possibly still requires some logic around updating patients

DEB-20

MattyJ007 committed:
Add OpenCR config update

OpenCR depends on files (or Env Vars) for configuration. To configure
OpenCR for this usecase we need a new config file that would overwrite
the generic data from the default client package settings.
Environment variables are not a viable option here as that would
require recreating the OpenCR container (instead of just restarting it)

By copying in the new file into the volumed directory the config will be
persisted and the container can just be restarted to get the new
settings.

DEB-20

MattyJ007 committed:
Improve the constant ids name

The mapping mediator currently cannot access data from request headers,
therefore we have hard coded in the request data from OpenCR until we
can add the mapping mediator feature.
The previous constant value did not match the format of the data very
well. This update string matches the expected content format

DEB-30

MattyJ007 committed:
Add input validation for patient FHIR resource

Enforce Covid Immunization IG requirements as far as possible on
incoming data before attempting to store it on FHIR

DEB-20

MattyJ007 committed:
Add update patient schema flow

CommCare will be sending through patient create messages as well as
updates to us. We need the appropriate flows to cater for these events

DEB-20

MattyJ007 committed:
Remove unused field 'id' from create schema

ID is only needed in the update input schema

DEB-20

MattyJ007 committed:
Add files for configuring this package

We need an OpenHIM channel setup as well as two mediator endpoints

DEB-20

MattyJ007 committed:
Bump the mapping mediator version

We require the latest features for this flow
Object pass through

DEB-20

MattyJ007 committed:
Add TODO about OpenCR config

This config is currently not easily implemented in Kubernetes. Once we
are in the kubernetes headspace this will be easier to configure and
test

DEB-20

de-laz reviewed (commented):
@MattyJ007 I'm happy with the changes but I am getting an error when trying to send the example patient to `http://localhost:5001/patient-registration`
![image](https://user-images.githubusercontent.com/3382449/110912531-b988b800-831c-11eb-8138-9fd4bf0161c7.png)

de-laz reviewed (approved):
@MattyJ007 DEB-20 LGTM :+1:
